packages/utils/src/utils/transpileAbi.ts

```diff
@@ -0,0 +1,152 @@
+    if (type) {
+      return type.typeId;
+    }
+
+    types.push({
+      typeId: types.length,
+      type: concreteType.type,
+      components: concreteType.type === '()' ? [] : parseComponents(concreteType.components),
+      concreteTypeId: id,
+      typeParameters: concreteType.typeParameters ?? null,
+    });
+
+    return types.length - 1;
+  }
+
+  function parseFunctionTypeArguments(concreteType) {
+    return (
+      concreteType.typeArguments?.map((cTypeId) => {
+        const self = findConcreteTypeById(cTypeId);
+        const type = !isNaN(cTypeId) ? cTypeId : finsertTypeIdByConcreteTypeId(cTypeId);
+        return {
+          name: '',
+          type,
+          // originalTypeId: cTypeId,
+          typeArguments: parseFunctionTypeArguments(self),
+        };
+      }) ?? null
+    );
+  }
+
+  function parseFunctioIO(concreteTypeId, name) {
+    const type = finsertTypeIdByConcreteTypeId(concreteTypeId);
+    const concrete = findConcreteTypeById(concreteTypeId);
+    return {
+      name: name ?? '',
+      type,
+      // concreteTypeId,
+      typeArguments: parseFunctionTypeArguments(concrete),
+    };
+  }
+
+  function parseComponents(components) {
+    return (
+      components?.map((component) => {
+        const { typeId, name, typeArguments } = component;
+        const type = !isNaN(typeId) ? typeId : finsertTypeIdByConcreteTypeId(typeId);
+        return {
+          name,
+          type,
+          // originalTypeId: typeId,
+          typeArguments: parseComponents(typeArguments),
+        };
+      }) ?? null
+    );
+  }
+
+  /**
+   * Transpiling
+   */
+
+  // 1. root level of metadata types
+  abi.metadataTypes.forEach((m) => {
+    const t = {
+      typeId: m.metadataTypeId,
+      type: m.type,
+      components: m.components ?? (m.type === '()' ? [] : null),
+      typeParameters: m.typeParameters ?? null,
+    };
+    types.push(t);
+  });
+
+  // 2. the metadata's components
+  types.forEach((t) => {
+    t.components = parseComponents(t.components);
+  });
+
+  // 3. functions inputs/outputs
+  const functions = abi.functions.map((fn) => {
+    const inputs = fn.inputs.map(({ concreteTypeId, name }) =>
+      parseFunctioIO(concreteTypeId, name)
+    );
+    const output = parseFunctioIO(fn.output, '');
+    return { ...fn, inputs, output };
+  });
+
+  // 4. configurables
+  const configurables = abi.configurables.map((conf) => ({
+    name: conf.name,
+    configurableType: parseFunctioIO(conf.concreteTypeId),
+    offset: conf.offset,
+  }));
+
+  // 5. loggedTypes
+  const loggedTypes = abi.loggedTypes.map((log) => ({
+    logId: log.logId,
+    loggedType: parseFunctioIO(log.concreteTypeId),
+  }));
+
+  // transpiled ABI
+  const transpiled = {
+    encoding: abi.encodingVersion,
+    types,
+    functions,
+    loggedTypes,
+    messagesTypes: abi.messagesTypes,
+    configurables,
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return transpiled as any;
```

Comment:
Can this now be cast as `JsonABI`?

Reply:
Yep, this script was initially a typeless `.mjs`.

However, the new ABI interface is in another PR and if we were to apply typings to the transpiler, it should go there:
 - https://github.com/FuelLabs/fuels-ts/pull/2862

Reply:
In that PR I just cast it wherever I use it:
```ts
const transpiled = transpileAbi(jsonAbi) as JsonAbi
```

The file has `// @ts-nocheck` in it and removing it brings about a lot of compilation errors. There wouldn't be any real benefit to fully typing it because we won't be using it anywhere else nor touching it except to delete it, so I'd just leave it as it is.

Reply:
Yep, fine by me.